builder/cloudstack/config.go

```diff
@@ -0,0 +1,309 @@
+	"github.com/mitchellh/packer/helper/communicator"
+	"github.com/mitchellh/packer/helper/config"
+	"github.com/mitchellh/packer/packer"
+	"github.com/mitchellh/packer/template/interpolate"
+	"github.com/xanzy/go-cloudstack/cloudstack"
+)
+
+// Config holds all the details needed to configure the builder.
+type Config struct {
+	common.PackerConfig `mapstructure:",squash"`
+	Comm                communicator.Config `mapstructure:",squash"`
+
+	APIURL       string `mapstructure:"api_url"`
+	APIKey       string `mapstructure:"api_key"`
+	SecretKey    string `mapstructure:"secret_key"`
+	AsyncTimeout int64  `mapstructure:"async_timeout"`
+	HTTPGetOnly  bool   `mapstructure:"http_get_only"`
+	SSLNoVerify  bool   `mapstructure:"ssl_no_verify"`
+
+	DiskOffering      string   `mapstructure:"disk_offering"`
+	DiskSize          int64    `mapstructure:"disk_size"`
+	CIDRList          []string `mapstructure:"cidr_list"`
+	Hypervisor        string   `mapstructure:"hypervisor"`
+	InstanceName      string   `mapstructure:"instance_name"`
+	Keypair           string   `mapstructure:"keypair"`
+	Network           string   `mapstructure:"network"`
+	Project           string   `mapstructure:"project"`
+	PublicIPAddress   string   `mapstructure:"public_ip_address"`
+	ServiceOffering   string   `mapstructure:"service_offering"`
+	SourceTemplate    string   `mapstructure:"source_template"`
+	SourceISO         string   `mapstructure:"source_iso"`
+	UserData          string   `mapstructure:"user_data"`
+	UseLocalIPAddress bool     `mapstructure:"use_local_ip_address"`
+	Zone              string   `mapstructure:"zone"`
+
+	TemplateName            string `mapstructure:"template_name"`
+	TemplateDisplayText     string `mapstructure:"template_display_text"`
+	TemplateOS              string `mapstructure:"template_os"`
+	TemplateFeatured        bool   `mapstructure:"template_featured"`
+	TemplatePublic          bool   `mapstructure:"template_public"`
+	TemplatePasswordEnabled bool   `mapstructure:"template_password_enabled"`
+	TemplateRequiresHVM     bool   `mapstructure:"template_requires_hvm"`
+	TemplateScalable        bool   `mapstructure:"template_scalable"`
+	TemplateTag             string `mapstructure:"template_tag"`
+
+	ctx            interpolate.Context
+	hostAddress    string // The host address used by the communicators.
+	instanceSource string // This can be either a template ID or an ISO ID.
+}
+
+// NewConfig parses and validates the given config.
+func NewConfig(raws ...interface{}) (*Config, error) {
+	c := new(Config)
+	err := config.Decode(c, &config.DecodeOpts{
+		Interpolate:        true,
+		InterpolateContext: &c.ctx,
+		InterpolateFilter: &interpolate.RenderFilter{
+			Exclude: []string{
+				"run_command",
+			},
+		},
+	}, raws...)
+	if err != nil {
+		return nil, err
+	}
+
+	var errs *packer.MultiError
+
+	// Set some defaults.
+	if c.AsyncTimeout == 0 {
+		c.AsyncTimeout = 1800
+	}
+
+	if c.Comm.SSHUsername == "" {
+		c.Comm.SSHUsername = "root"
```

Comment:
I fully agree, but don't want to make that change in this PR. I would say a change like that deserves it's own PR... Agree?